File: walkthrough/scripts/generate-voiceover.ts
/**
 * Generate narration for the assessment ad and write a manifest the video reads.
 *   ELEVENLABS_API_KEY=...  npm run voiceover     (or OPENAI_API_KEY)
 *
 * Produces one clip per beat (public/narration/assessment/<id>.mp3), measures
 * each, and writes src/assessment.narration.json. The ad then stretches each
 * footage segment to its spoken line. Run where the network reaches the TTS host.
 */
import { writeFile, mkdir } from "node:fs/promises";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { NARRATION } from "../src/narration-script";

const here = dirname(fileURLToPath(import.meta.url));
const root = join(here, "..");
const audioDir = join(root, "public", "narration", "assessment");
const manifestPath = join(root, "src", "assessment.narration.json");
const TAIL_PAD = 0.45; // seconds of breathing room after each line

// MP3 duration straight from the bytes (ElevenLabs returns CBR MPEG-1 L3 128k).
const BR = [0, 32, 40, 48, 56, 64, 80, 96, 112, 128, 160, 192, 224, 256, 320, 0];
function mp3Seconds(buf: Buffer): number {
  let i = 0;
  if (buf.length > 10 && buf.toString("latin1", 0, 3) === "ID3") {
    i = 10 + (((buf[6] & 0x7f) << 21) | ((buf[7] & 0x7f) << 14) | ((buf[8] & 0x7f) << 7) | (buf[9] & 0x7f));
  }
  while (i < buf.length - 4 && !(buf[i] === 0xff && (buf[i + 1] & 0xe0) === 0xe0)) i++;
  const kbps = BR[(buf[i + 2] >> 4) & 0x0f] || 128;
  return ((buf.length - i) * 8) / (kbps * 1000);
}

async function proxy() {
  const p = process.env.HTTPS_PROXY || process.env.https_proxy;
  if (!p) return;
  const { ProxyAgent, setGlobalDispatcher } = await import("undici");
  setGlobalDispatcher(new ProxyAgent(p));
}

async function eleven(text: string): Promise<Buffer> {
  const voice = process.env.ELEVENLABS_VOICE_ID || "JBFqnCBsd6RMkjVDRZzb";
  const res = await fetch(`https://api.elevenlabs.io/v1/text-to-speech/${voice}?output_format=mp3_44100_128`, {
    method: "POST",
    headers: { "xi-api-key": process.env.ELEVENLABS_API_KEY as string, "content-type": "application/json" },
    body: JSON.stringify({ text, model_id: process.env.ELEVENLABS_MODEL || "eleven_multilingual_v2", voice_settings: { stability: 0.5, similarity_boost: 0.75 } }),
  });
  if (!res.ok) throw new Error(`ElevenLabs ${res.status}: ${await res.text()}`);
  return Buffer.from(await res.arrayBuffer());
}
async function openai(text: string): Promise<Buffer> {
  const res = await fetch("https://api.openai.com/v1/audio/speech", {
    method: "POST",
    headers: { authorization: `Bearer ${process.env.OPENAI_API_KEY}`, "content-type": "application/json" },
    body: JSON.stringify({ model: process.env.OPENAI_TTS_MODEL || "gpt-4o-mini-tts", voice: process.env.OPENAI_TTS_VOICE || "onyx", input: text, response_format: "mp3" }),
  });
  if (!res.ok) throw new Error(`OpenAI ${res.status}: ${await res.text()}`);
  return Buffer.from(await res.arrayBuffer());
}

async function main() {
  const provider = process.env.ELEVENLABS_API_KEY ? "elevenlabs" : process.env.OPENAI_API_KEY ? "openai" : null;
  if (!provider) { console.error("Set ELEVENLABS_API_KEY or OPENAI_API_KEY first."); process.exit(1); }
  console.log(`· provider: ${provider}`);
  await proxy();
  await mkdir(audioDir, { recursive: true });
  const speak = provider === "elevenlabs" ? eleven : openai;

  const sections: Record<string, { seconds: number; audio: string }> = {};
  for (const part of NARRATION) {
    process.stdout.write(`  ${part.id} … `);
    const mp3 = await speak(part.text);
    await writeFile(join(audioDir, `${part.id}.mp3`), mp3);
    const seconds = Number((mp3Seconds(mp3) + TAIL_PAD).toFixed(2));
    sections[part.id] = { seconds, audio: `narration/assessment/${part.id}.mp3` };
    console.log(`${seconds}s`);
  }
  await writeFile(manifestPath, JSON.stringify({ enabled: true, provider, voiceId: process.env.ELEVENLABS_VOICE_ID || process.env.OPENAI_TTS_VOICE || "default", sections }, null, 2) + "\n");
  console.log("\n✓ narration written. Now: npm run render-ad");
}
main().catch((e) => { console.error("\n✗", e.message); process.exit(1); });

File: walkthrough/scripts/generate_voiceover.py
#!/usr/bin/env python3
"""
Generate the assessment-ad narration with ElevenLabs (or OpenAI), and write the
manifest the Remotion video reads. Pure standard library — no pip installs.

USAGE (run from the `walkthrough/` folder):
    export ELEVENLABS_API_KEY=sk_...            # or: export OPENAI_API_KEY=sk-...
    # optional voice override:
    # export ELEVENLABS_VOICE_ID=JBFqnCBsd6RMkjVDRZzb
    python3 scripts/generate_voiceover.py

Then render the voiced ad:
    npm run render-ad        # -> out/assessment-ad.mp4
"""
import json
import os
import sys
import urllib.request
import urllib.error

# ----- the script: one line per beat (edit the words here) -------------------
NARRATION = [
    ("open",      "Most banks know AI matters. Almost none know where they actually stand."),
    ("questions", "So we made it simple. Twelve plain-language questions — about three minutes, no jargon."),
    ("form",      "Add your name and where you bank, and the result is yours."),
    ("report",    "And you get a real report: a score, your biggest gap, and a concrete first move you can make on Monday."),
    ("close",     "The AI Banking Institute. Find your starting point — free."),
]

TAIL_PAD = 0.45  # seconds of breathing room added after each line

# ----- paths (relative to where you run it, i.e. the walkthrough/ folder) -----
ROOT = os.getcwd()
AUDIO_DIR = os.path.join(ROOT, "public", "narration", "assessment")
MANIFEST = os.path.join(ROOT, "src", "assessment.narration.json")

# ----- MP3 duration straight from the bytes (ElevenLabs returns CBR 128k) -----
_BITRATES = [0, 32, 40, 48, 56, 64, 80, 96, 112, 128, 160, 192, 224, 256, 320, 0]


def mp3_seconds(buf: bytes) -> float:
    i = 0
    if len(buf) > 10 and buf[0:3] == b"ID3":
        i = 10 + (((buf[6] & 0x7F) << 21) | ((buf[7] & 0x7F) << 14) |
                  ((buf[8] & 0x7F) << 7) | (buf[9] & 0x7F))
    while i < len(buf) - 4 and not (buf[i] == 0xFF and (buf[i + 1] & 0xE0) == 0xE0):
        i += 1
    kbps = _BITRATES[(buf[i + 2] >> 4) & 0x0F] or 128
    return (len(buf) - i) * 8 / (kbps * 1000)


def _post(url: str, headers: dict, body: dict) -> bytes:
    req = urllib.request.Request(url, data=json.dumps(body).encode("utf-8"),
                                 headers=headers, method="POST")
    try:
        with urllib.request.urlopen(req) as r:
            return r.read()
    except urllib.error.HTTPError as e:
        raise SystemExit(f"\n✗ {url.split('/')[2]} {e.code}: {e.read().decode('utf-8', 'replace')[:300]}")


def tts_elevenlabs(text: str) -> bytes:
    voice = os.environ.get("ELEVENLABS_VOICE_ID", "JBFqnCBsd6RMkjVDRZzb")
    model = os.environ.get("ELEVENLABS_MODEL", "eleven_multilingual_v2")
    url = f"https://api.elevenlabs.io/v1/text-to-speech/{voice}?output_format=mp3_44100_128"
    headers = {"xi-api-key": os.environ["ELEVENLABS_API_KEY"], "content-type": "application/json"}
    body = {"text": text, "model_id": model,
            "voice_settings": {"stability": 0.5, "similarity_boost": 0.75}}
    return _post(url, headers, body)


def tts_openai(text: str) -> bytes:
    url = "https://api.openai.com/v1/audio/speech"
    headers = {"authorization": f"Bearer {os.environ['OPENAI_API_KEY']}", "content-type": "application/json"}
    body = {"model": os.environ.get("OPENAI_TTS_MODEL", "gpt-4o-mini-tts"),
            "voice": os.environ.get("OPENAI_TTS_VOICE", "onyx"),
            "input": text, "response_format": "mp3"}
    return _post(url, headers, body)


def main() -> None:
    if os.environ.get("ELEVENLABS_API_KEY"):
        provider, speak = "elevenlabs", tts_elevenlabs
    elif os.environ.get("OPENAI_API_KEY"):
        provider, speak = "openai", tts_openai
    else:
        raise SystemExit("Set ELEVENLABS_API_KEY (or OPENAI_API_KEY) first.")

    if not os.path.isdir(os.path.join(ROOT, "src")) or not os.path.isdir(os.path.join(ROOT, "public")):
        raise SystemExit("Run this from the walkthrough/ folder (it needs ./src and ./public).")

    print(f"· provider: {provider}")
    os.makedirs(AUDIO_DIR, exist_ok=True)

    sections = {}
    for beat_id, text in NARRATION:
        sys.stdout.write(f"  {beat_id} … "); sys.stdout.flush()
        audio = speak(text)
        with open(os.path.join(AUDIO_DIR, f"{beat_id}.mp3"), "wb") as fh:
            fh.write(audio)
        seconds = round(mp3_seconds(audio) + TAIL_PAD, 2)
        sections[beat_id] = {"seconds": seconds, "audio": f"narration/assessment/{beat_id}.mp3"}
        print(f"{seconds}s")

    manifest = {
        "enabled": True,
        "provider": provider,
        "voiceId": os.environ.get("ELEVENLABS_VOICE_ID", os.environ.get("OPENAI_TTS_VOICE", "default")),
        "sections": sections,
    }
    with open(MANIFEST, "w") as fh:
        fh.write(json.dumps(manifest, indent=2) + "\n")

    print(f"\n✓ wrote {len(sections)} clips + manifest. Now run:  npm run render-ad")


if __name__ == "__main__":
    main()
